### ISSUE TITLE
python statsmodel WLS memory/gc “leak” at n=1024 boundary

### ISSUE THREAD

**Issue by pycy** [statsmodels/statsmodels#2428]
I run jobs with loops that call statsmodels.api.WLS several times per loop. It appears that statsmodels (version 0.6.1) refuses to give up memory despite my best attempts and ends up taking up all my memory (32Gb). I wrote the following example code to demonstrate the problem.

The interesting thing is that there is no problem if the dimension of the problem is within 1024 (n =< 1024). **However, if ~~n >= 1025~~  2^10 < n <  2^17 +1, then this script's memory usage increases linearly without bound.** Looks like an error in statsmodel.api but it could be my poor understanding of python:
(also submitted to https://stackoverflow.com/questions/30543287/python-statsmodel-wls-memory-gc-leak-at-n-1024-boundary)

```
#!/usr/bin/python
import resource
import numpy as np
import statsmodels.api as sm

if __name__ == "__main__":
    n = 1025
    mu, sigma = 0.0, 1.0
    i = 0
    print('i,memory')
    while True:
        x = np.random.normal(mu, sigma, n)
        y = np.random.normal(mu, sigma, n)
        w = np.random.random(n)
        w /= w.sum()
        regr1 = sm.WLS(y, x, w)
        print(`i` + ',' + `resource.getrusage(resource.RUSAGE_SELF).ru_maxrss`)
        i+=1
```

Here's an example memory usage for n=1000 and n=10000 respectively:
![bla](https://cloud.githubusercontent.com/assets/9191834/7896018/3cdf2512-06dd-11e5-91bc-2c7f1690c8fd.jpg)

The first 10 values of n=1024 and n=1025:
```
i	N_1024	N_1025
0	60680	60812
1	60680	60848
2	60816	60988
3	60816	60996
4	60816	61016
5	60816	61028
6	60816	61040
7	60816	61064
8	60816	61080
9	60816	61084
10	60816	61092
```

**Comment by rgommers** [comment]
I can't reproduce this with:

- statsmodels master
- scipy master
- numpy 1.9.1
- python 2.7.6
- 32-bit Ubuntu 14.04

What's your version of all of the above?

**Comment by pycy** [comment]
* statsmodels 0.6.1
* scipy 0.15.1
* numpy 1.7.1
* python 2.7.5 (default, Jun 17 2014, 18:11:42) 
* 64bit Centos 7.0

**Comment by pycy** [comment]
@rgommers -- I upgraded to numpy 1.9.1 and the problem automagically disappeared.

**Comment by rgommers** [comment]
Nice, thanks. It almost had to be a numpy or scipy issue somewhere.

**Comment by josef-pkt** [comment]
Thanks for reporting and checking.

I think it's still strange even with numpy 1.7. (I don't have a numpy 1.7 installed right now to check.)
Loops that create a WLS instance and fit it, are used both in GLM and in RLM and I never heard of a memory problem (since early versions of statsmodels).

In early versions of statsmodels we had a reference cycle after calling `fit` that delayed garbage collection and was keeping large numpy arrays in storage for quite some time.
Creating the WLS model just does the initial data handling and calculates the weighted x and y but shouldn't do anything/much else.

We don't have unit tests for memory usage and garbage collection, so there is no guarantee that another reference cycle doesn't sneak in, but we try to be careful to avoid those.

**Comment by pycy** [comment]
@josef-pkt happy to help test different situations for you if you think of anything although I've only been using python for about a month so my knowledge of how python works is very basic.  My use case is about 5000 runs of a (20000 x 50) WLS.  I could replicate the same problem using OLS and GLS (a quick look at the source shows they are the same thing so maybe that's not very useful).  

Also, before upgrading to numpy 1.9.1, I also had situations where the regression just randomly hung.  

Hope that's useful to someone.

**Comment by josef-pkt** [comment]
@pycy Is it possible that you have `nans` or `infs` in the data?
We are not currently checking this by default. (AFAIK, we had an implicit check that got removed when we streamlined the linear algebra to reduce extra, redundant or duplicate work.) 
It could also be that with a newer version of numpy it uses also a newer version of BLAS/LAPACK, which in some versions could hang or even segfault if there are nans or infs. (But then you would see an exception raised by BLAS or LAPACK.)

However, all the linear algebra should now be in `fit` only (*), and just creating a model shouldn't do much.


(*) technical detail: it's not completely true, there is one case where we check if there is a constant in the model which uses linear algebra if there is no explicit constant.


-----------

In general:
(aside from our ability to mess up the python garbage collection)

We only use cython generated c code in nonparametrics and tsa, so for most parts it is essentially impossible that statsmodels creates a memory leak.
However, we make extensive use of numpy and scipy and the underlying linear algebra libraries and other fortran and C libraries. Any problems there could cause problems for us, and we usually follow up on those. Either we report them to numpy or scipy, or we work around, or we just "warn".

In this case, it looks like it's already fixed with your numpy 1.9 version, so it's not really of any priority to figure out what caused it and who fixed it. But it's always useful to keep track of where and when problems might show up for users.

**Comment by pycy** [comment]
@josef-pkt, the memory leak exists even in the example code above, which should not have any nan's/inf's.  My gut feeling (not substantiated at all) is perhaps something to do float vs double or 32 bit / 64 bit implementations mixed together but now I'm really starting to get voodoo.
